File: python/152. Maximum Product Subarray.py
class Solution(object):
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums:
            return 0
        maxval = minval = res = nums[0]
        for i in range(1, len(nums)):
            temp = maxval
            maxval = max(nums[i], maxval * nums[i], minval * nums[i])
            minval = min(nums[i], temp * nums[i], minval * nums[i])
            res = max(maxval, res)
        return res

File: cpp/151-200/152. Maximum Product Subarray.cc
class Solution {
public:
    int maxProduct(vector<int>& nums) {
    	if(nums.empty())
    		return 0;
    	int res = nums[0];
    	int imax = res, imin = res;
    	for(int i = 1; i < nums.size(); i++){
    		if(nums[i] < 0)
    			swap(imax, imin);
    		imax = max(nums[i], imax * nums[i]);
    		imin = min(nums[i], imin * nums[i]);
    		res = max(res, imax);
    	}
    	return res;
    }
};